File: 1663-detect-cycles-in-2d-grid/detect-cycles-in-2d-grid.cpp
class Solution {
public:
    bool containsCycle(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<int> parent(m * n);
        vector<int> rank(m * n, 0);

        // initialize parent array
        iota(parent.begin(), parent.end(), 0);

        // lambda for path compression
        auto find = [&](auto& self, int i) -> int {
            if (parent[i] == i)
                return i;
            return parent[i] = self(self, parent[i]);
        };

        for (int r = 0; r < m; ++r) {
            for (int c = 0; c < n; ++c) {
                // checking right and down
                int neighbors[2][2] = {{r, c + 1}, {r + 1, c}};
                for (auto& neighbor : neighbors) {
                    int nr = neighbor[0], nc = neighbor[1];

                    if (nr < m && nc < n && grid[r][c] == grid[nr][nc]) {
                        int u = r * n + c;
                        int v = nr * n + nc;

                        int rootU = find(find, u);
                        int rootV = find(find, v);

                        if (rootU == rootV)
                            return true;

                        // union by rank
                        if (rank[rootU] > rank[rootV])
                            parent[rootV] = rootU;
                        else if (rank[rootU] < rank[rootV])
                            parent[rootU] = rootV;
                        else {
                            parent[rootU] = rootV;
                            rank[rootV]++;
                        }
                    }
                }
            }
        }
        return false;
    }
};

File: 1663-detect-cycles-in-2d-grid/detect-cycles-in-2d-grid.js
var containsCycle = function (grid) {
    const m = grid.length;
    const n = grid[0].length;
    const parent = new Array(m * n);
    const rank = new Array(m * n).fill(0);

    // Initialize parent array
    for (let i = 0; i < m * n; i++) parent[i] = i;

    const find = (i) => {
        if (parent[i] === i) return i;
        parent[i] = find(parent[i]); // path compression
        return parent[i];
    };

    for (let r = 0; r < m; r++) {
        for (let c = 0; c < n; c++) {
            const directions = [[r, c + 1], [r + 1, c]];

            for (const [nr, nc] of directions) {
                if (nr < m && nc < n && grid[r][c] === grid[nr][nc]) {
                    const u = r * n + c;
                    const v = nr * n + nc;

                    const rootU = find(u);
                    const rootV = find(v);

                    if (rootU === rootV) return true;

                    // Union by rank
                    if (rank[rootU] > rank[rootV]) {
                        parent[rootV] = rootU;
                    } else if (rank[rootU] < rank[rootV]) {
                        parent[rootU] = rootV;
                    } else {
                        parent[rootU] = rootV;
                        rank[rootV]++;
                    }
                }
            }
        }
    }
    return false;
};